### DISCUSSION TITLE
Please make Constructors DI Friendly

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#4156] [Language Ideas]
Let's say you have a base class like so:

```
    public class UiExtension<TInterface> : Extension
        where TInterface:class 
        
        {
        public WpfApplicationExtension Application { get; private set; }
        public TInterface? Interface { get; private set; }
      
        public UiExtension(WpfApplicationExtension Application, LogProvider LogProvider) : base(LogProvider) {
            this.Application = Application;
        }
}
```

And you want to use a derived version of it.

Every time you derive from it you have to re-specify all of the constructors you want.  For example:

```
    public class MainWindowExtension : UiExtension<MainWindow> {
        public MainWindowExtension(WpfApplicationExtension Application, LogProvider LogProvider) : base(Application, LogProvider) {
        
        }

        protected override Task InitializeInterfaceAsync(MainWindow Interface) {
            Interface.Show();

            return Task.CompletedTask;
        }

    }
```

This creates a lot of repetitive code and makes refactoring suck since adding a parameter to a base class requres that you refactor every single class that derives from it.

With regular methods, in a scenario like that, you only have to refactor explicit overrides.

I propose that if no constructors are specified on a class, that constructors be synthensized that match the parameters of the immediate base class.

**Comment by CyrusNajmabadi** [comment]
> I propose that if no constructors are specified on a class, that constructors be synthensized that match the parameters of the immediate base class.

Similar to requests to restrict the language, requests to generate purely additive code are very well suited to just using a Source-Generator.  This would be a very easy one to write IMO.
